src/ir/datalog/input_relation_fact.h

```diff
@@ -0,0 +1,54 @@
+//-----------------------------------------------------------------------------
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//----------------------------------------------------------------------------
+
+#ifndef SRC_IR_DATALOG_INPUT_RELATION_FACT_H_
+#define SRC_IR_DATALOG_INPUT_RELATION_FACT_H_
+
+#include <utility>
+
+#include "absl/strings/str_format.h"
+#include "absl/strings/str_join.h"
+#include "src/ir/datalog/value.h"
+
+namespace raksha::ir::datalog {
+
+// A fact that is built to be fed into an input relation for Souffle.
+// Although the `Rule` class can also express a `Fact` by having no RHS, this
+// is explicitly to prepare input to a Souffle program. You can think of the
+// difference between this `Fact` and a `Rule` fact as a difference between
+// data and code.
```

Comment:
If this is data, shouldn't this be a `datalog::Value`? It looks like a record might be sufficient for what you are trying to do?

I don't clearly understand  what is the advantage of this over `Rule`. e.g., `ToDatalogString` generates the same representation as a `Rule`? What do we gain by having a separate class?

Reply:
Is the difference that the arguments should be constant terms?

Reply:
Yes, sort of. For an `InputRelationFact`, we have strong ideas about what kind of static type each of the arguments to that `InputRelationFact` should be. `InputRelationFact` therefore keeps around static type information via templates, which allows the C++ type checker to catch us if we try to do something silly like get our `AttributeList` and `OperationList` backwards. Using full types instead of string info makes certain classes of errors just impossible, such as missing a closing `]` in a relation definition.

Because the authorization logic translator has to deal with custom relations that a user may introduce, it must have a dynamic representation of even `Rule`s that happen to be facts. You could imagine it augmenting its AST with these static type annotations for "core" relations used by the authorization logic though, so it does make sense to place these classes here in hopes of future sharing. But the Souffle backend for the dataflow graph IR cannot use the existing `Rule` class without losing a great deal of valuable type information.

Reply:
I think this is more closely related to `Predicate` than `Rule`. We might want to reconcile these two at some point. I like the type safety, but I expect this to add some friction due to lack of flexibility. Let us run with this see how things go. 

I might be OK calling this simply a `Fact` as well.

Reply:
Yes, I see that. Ideally, I'd like `InputRelationFact` to be a static-typed wrapper around the less typed `Predicate`. Unfortunately, `Predicate` is currently too unstructured to be used this way, containing only a `vector` of `std::string`s. If we worked to integrate them, I'd like the arguments of `Predicate` to be represented as a `vector` of `Value`s similar to the ones I've created (or similar type-erased objects).